### ISSUE TITLE
Unable to mount hostPath file volume to a directory mounted by a secret volume

### ISSUE THREAD

**Issue by robbiezhang** [kubernetes/kubernetes#62417]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
Apply the following yaml in a k8s 1.8.10+ cluster, and the pod gets into "RunContainerError" state.
The event shows:

```
Error: failed to start container "nginx": Error response from daemon: oci runtime error: container_linux.go:247: starting container process caused "process_linux.go:359: container init caused \"rootfs_linux.go:54: mounting \\\"/var/log/waagent.log\\\" to rootfs \\\"/var/lib/docker/overlay2/4407304a98d14267ac820acfbccf078699227e165b6c889e8d100cf2d1dd73b9/merged\\\" at \\\"/var/lib/docker/overlay2/4407304a98d14267ac820acfbccf078699227e165b6c889e8d100cf2d1dd73b9/merged/var/log/lala/aaa.log\\\" caused \\\"not a directory\\\"\""
: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type
```
Here is the yaml:
```
apiVersion: v1
kind: Secret
metadata:
  name: nginx-secret
type: Opaque
data:
  aaa: TUlTU0lORw==

---

apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - image: nginx
    name: nginx
    ports:
    - containerPort: 80
      name: http
      protocol: TCP
    volumeMounts:
    - mountPath: /var/log/lala
      name: secret-dir
    - mountPath: /var/log/lala/aaa.log
      name: host-file
  volumes:
  - name: secret-dir
    secret:
      secretName: nginx-secret
  - name: host-file
    hostPath:
      path: /var/log/waagent.log
      type: File
```
Note: It works in k8s 1.8.7 or lower version.

**What you expected to happen**:
The pod is created successfully.

**How to reproduce it (as minimally and precisely as possible)**:
Replace the hostPath.path in the yaml above and apply to a k8s cluster 1.8.10+

**Anything else we need to know?**:
It works in k8s 1.8.7 or lower.
If the directory is mounted by an emptyDir volume, the pod can start without any issue.

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.1", GitCommit:"3a1c9449a956b6026f075fa3134ff92f7d55f812", GitTreeState:"clean", BuildDate:"2018-01-04T11:52:23Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.10", GitCommit:"044cd262c40234014f01b40ed7b9d09adbafe9b1", GitTreeState:"clean", BuildDate:"2018-03-19T17:44:09Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
- Cloud provider or hardware configuration:
Azure (AKS)
- OS (e.g. from /etc/os-release):
NAME="Ubuntu"
VERSION="16.04.3 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.3 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
- Kernel (e.g. `uname -a`):
Linux aks-nodepool1-27738643-1 4.13.0-1011-azure #14-Ubuntu SMP Thu Feb 15 16:15:39 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
Azure CLI (az aks create)
- Others:

**Comment by andyzhangx** [comment]
@robbiezhang It's an expected error due to security fix from v1.8.9: 
Fixed versions:
	• Fixed in v1.7.14 by #60516
	• Fixed in v1.8.9 by #60515
	• Fixed in v1.9.4 by #60258
	• Fixed in master by #58720 (included in v1.10.0-beta.1 and up, will be in v1.10.0)
Fix impact:
Secret, configMap, downwardAPI and projected volumes will be mounted as read-only volumes. Applications that attempt to write to these volumes will receive read-only filesystem errors. Previously, applications were allowed to make changes to these volumes, but those changes were reverted at an arbitrary interval by the system. Applications should be re-configured to write derived files to another location.
From <https://github.com/kubernetes/kubernetes/issues/60814> 

Let me explain why it works before v1.8.9:
1. `host-file` will be mounted as `/var/log/lala/aaa.log` in container, at this time, `/var/log/lala` is a dir
2. `secret-dir` will overwrite `/var/log/lala` dir as secrect, this time `/var/log/lala` becomes `tmpfs`
3. In the container init process, hostpath volume mapping could overwrite `/var/log/lala` since it's readwrite before v1.8.9, while starting from v1.8.9, `/var/log/lala` is readonly, so it fails.

Anyway, your volumeMounts config is conflicted, change as below would work:
```
    volumeMounts:
    - mountPath: /var/log/lala2
      name: secret-dir
    - mountPath: /var/log/lala/aaa.log
      name: host-file
```

**Comment by andyzhangx** [comment]
I will close this issue, let me know if you have any question, thanks.
/close
